enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+before instructing the cluster to operate on the changed configuration,
+so that I can check that the features I have tuned are converted correctly to the new configuration format.
+
+#### Story 3
+
+As a cluster administrator, I would like to test Cluster API before rolling it out to my production clusters, so that I can build confidence in the new APIs.
+To do this, I would like to be able to add Cluster API MachineSets to add new Machines to my Cluster without having to have an equivalent Machine API MachineSet to match it.
+
+#### Story 4
+
+As a developer of Machine related APIs, I would like to be able to gather feedback on Machine conversion layers before forcing new APIs onto users.
+By gathering feedback, we can ensure the conversion is working, and that we are not breaking existing users when they are upgraded into newer versions
+of OpenShift that rely on Cluster API.
+
+### Goals
+
+* Allow users to migrate Machines between Machine API and Cluster API to test the new API
+* Allow users a choice (initially[^3]) of when they want to migrate
+* Eventually[^3] remove the Machine API controllers and leverage the migration layer to convert Machine API machines to Cluster API machines
+* Allow the Cluster Autoscaler, MachineHealthCheck and ControlPlaneMachineSet to continue to work during a semi-migrated cluster state
+* Ensure that every Machine API resource is translated to a Cluster API resource to allow users to test the new API
+* Allow access to newer features of Cluster API that do not exist in Machine API
+
+[^3]: In the first iteration of this project, the goal is to allow users to migrate individual resources.
+In later iterations, the plan is to remove the Machine API controllers. The linked goals will be executed in separate releases.
+
+### Non-Goals
+
+* Setting up or configuring any part of Cluster API, this is handled in a separate enhancement
+* Modifying any of the existing deployments of Machine API within clusters
+* Deleting or modifying any existing resources within clusters (until a user chooses to do so)
+* Force creation of new Machine API machines when new Cluster API MachineSets are being created[^4]
+* Port any feature from Cluster API into Machine API
+
+[^4]: We want to promote, over time, a reduction in the number of Machine API Machines, as such,
+any resources created directly in Cluster API, will not be reflected into Machine API, unless they have an owner,
+and, that owner is already reflected into Machine API.
+
+## Proposal
+
+We will implement a two-way sync controller that synchronises Machines and related resources between the Machine API and Cluster API equivalents.
+
+Using a new field on the Machine API resource, users will be able to choose which API is authoritative.
+Any changes to the non-authoritative resource will be overwritten by the sync controller to match the authoritative resource.
```

Comment:
If the field is on Machine API, how does Cluster API's controllers know when to filter out? A label selector?

Reply:
the doc says resource, the comment says field.  I'm not clear which we're discussing here.

Reply:
We are talking about both.

For each entity (where we count a MAPI and CAPI machine representing the same cloud infrastructure as one entity), we want one API field that determines which of the resources (MAPI machine or CAPI machine) is authoritative.

For simplicity sake, we are saying that this field should live on the MAPI resource (MAPI machine). But if the field on the MAPI machine says that the `authoritativeAPI: CAPI`, then the CAPI resource (CAPI machine) is referred to as the authoritative resource, even though the API to dictate that, is now on the non-authoritative resource.

Every spec field in this case, from the CAPI resource, would be copied and stomp the spec of the MAPI resource, except the `authoritativeAPI` field, which wouldn't exist in the CAPI resource spec